### PR TITLE
[AXLE-31] Publish session activities in RabbitMQ

### PR DESCRIPTION
**Pull request by iliyan-georgiev-ph**

### CODE REVIEW

File: src/Axle/Hubs/SessionHub.cs [hunk 1]
```diff
@@ -37,17 +40,39 @@ public override async Task OnConnectedAsync()
...
+            if (string.IsNullOrWhiteSpace(accountId))
+            {
+                Log.Information($"User {sub} tried to connect without specifying account ID.");
+                this.Context.Abort();
+                return;
+            }
+
+            var accountOwnerId = await this.accountsService.GetAccountOwnerUserId(accountId);
+
+            if (sub != accountOwnerId)
```
Comment (Waqas-Zamir) [inline src/Axle/Hubs/SessionHub.cs]: we need to check this by account owner authorization policy, as support users will have on behalf access as well

Comment (Waqas-Zamir) [inline src/Axle/Hubs/SessionHub.cs]: we need to check this by account owner authorization policy, as support users will have on behalf access as well

File: src/Axle/Hubs/SessionHub.cs [hunk 2]
```diff
@@ -37,17 +40,39 @@ public override async Task OnConnectedAsync()
...
-            await this.sessionLifecycleService.OpenConnection(this.Context.ConnectionId, sub, clientId, token);
 
-            Log.Information($"New connection established (ID: {this.Context.ConnectionId}).");
-            this.connectionRepository.Add(this.Context.ConnectionId, this.Context);
+            var query = this.Context.GetHttpContext().Request.Query;
+
+            var token = query["access_token"];
+            var accountId = query["account_id"];
+
+            if (string.IsNullOrWhiteSpace(accountId))
```
Comment (Waqas-Zamir) [inline src/Axle/Hubs/SessionHub.cs]: what about support users

File: src/Samples/Axle.Samples.SinglePageApplication/wwwroot/index.html [hunk 3]
```diff
@@ -227,7 +227,7 @@ <h1>Single Page Application</h1>
             mgr.getUser().then(function (user) {
                 if (user) {
                     connection = new signalR.HubConnectionBuilder()
-                        .withUrl("http://localhost:5012/session",
+                        .withUrl("http://localhost:5012/session?account_id=YOUR_ACCOUNT_HERE",
```
Comment (Waqas-Zamir) [inline src/Samples/Axle.Samples.SinglePageApplication/wwwroot/index.html]: may be you get this from some textbox

**Comment by Waqas-Zamir** [comment]
why you are targeting to master?

Also rebase on top of dev branch to get the latest changes from it

**Comment by iliyan-georgiev-ph** [comment]
> why you are targeting to master?
> 
> Also rebase on top of dev branch to get the latest changes from it

I probably didn't see that the PR was targeting master. I've changed it. I'll rebase too

File: src/Axle/Hubs/SessionHub.cs [hunk 4]
```diff
@@ -16,6 +17,7 @@ namespace Axle.Hubs
     using Serilog;
 
     [Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme)]
+    [AccountOwnerOrSupport]
```
Comment (Waqas-Zamir) [inline src/Axle/Hubs/SessionHub.cs]: you don't need to create or put custom attribute here,
just pass the policy name to Authorize attribute above, like
`[Authorize(AuthenticationSchemes = JwtBearerDefaults.AuthenticationScheme, Policy = AuthorizationPolicies.AccountOwnerOrSupport)]`

File: src/Axle/Hubs/SessionHub.cs [hunk 5]
```diff
@@ -35,19 +37,27 @@ public class SessionHub : Hub
...
-            await this.sessionLifecycleService.OpenConnection(this.Context.ConnectionId, sub, clientId, token);
 
-            Log.Information($"New connection established (ID: {this.Context.ConnectionId}).");
-            this.connectionRepository.Add(this.Context.ConnectionId, this.Context);
+            var query = this.Context.GetHttpContext().Request.Query;
+
+            var token = query["access_token"];
+            var accountId = query["account_id"];
+
+            var isSupportUser = AccountOwnerOrSupportHandler.IsSupportUser(accountId, this.Context.User);
```
Comment (Waqas-Zamir) [inline src/Axle/Hubs/SessionHub.cs]: it is not a good idea to use method from AuthorizationHandler in hub. SRP violation

File: src/Axle/Authorization/AccountOwnerOrSupportHandler.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        private readonly IHttpContextAccessor contextAccessor;
+        private readonly IAccountsService accountsService;
+
+        public AccountOwnerOrSupportHandler(IHttpContextAccessor contextAccessor, IAccountsService accountsService)
+        {
+            this.contextAccessor = contextAccessor;
+            this.accountsService = accountsService;
+        }
+
+        public static bool IsSupportUser(string accountId, ClaimsPrincipal user)
```
Comment (Waqas-Zamir) [inline src/Axle/Authorization/AccountOwnerOrSupportHandler.cs]: SRP violation, better to create extension method on `ClaimsPrincipal` (user)

File: src/Axle/Caches/IAccountsCache.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+namespace Axle.Caches
+{
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    public interface IAccountsCache
+    {
+        Task<List<string>> GetAccountIds(string clientId);
+
+        Task<string> GetClientId(string accountId);
```
Comment (Waqas-Zamir) [inline src/Axle/Caches/IAccountsCache.cs]: I couldn't find where this method is being used

**Pull request merged**